### PR TITLE
security: enforce strict input validation and payload bounds in tokens.Service (Fixes #1608)

### LINKED ISSUE #1608: Enforce Strict Input Validation and Payload Bounds on Token Store Writes [MED]
**Description**
Even after a delivery event has been authenticated, the finality listener must not blindly commit its content to the Token Store. Before invoking `StoreToken` or `DeleteTokens`, enforce explicit bounds and well-formedness checks on the incoming data: maximum payload size for raw token bytes (`ledger`, `ledger_metadata`), maximum number of token outputs per transaction, maximum batch size for bulk delete operations, and rejection of structurally invalid or oversized `wallet_id` / `owner_raw` / `issuer_raw` fields. These checks should be applied as a validation stage in the `tokens.Service.Append` call chain, before any SQL statement is issued, ensuring that even a legitimate but misbehaving peer cannot cause unbounded storage or CPU consumption in `tokendb`.

**Short Summary**
Resource exhaustion / DoS attack. Malicious users can create toxic transaction that tokenSDK will submit to the ledger and receive back finality events that cause this issue. A solution can be limiting the message size, metadata size, etc. and some validation checks.
To address this issue we need:
(1) To prevent the Ledger from sending large messages in finality we need to limit the protobuf messages. This needs to be updated in the documentation accordingly.
(2) When Alice transfers to Bob, bob needs to validate that Alice's token is not too big.

### PR DESCRIPTION
**Pull request by SurbhiAgarwal1**
## Description
This PR implements strict input validation and payload bounds within the `tokens.Service.Append` method to protect the system against resource exhaustion and potential Denial of Service (DoS) attacks from malformed or oversized finality data.

## Changes
- **ValidationConfig**: Introduced a configurable validation layer with safe default limits:
    - `MaxTokenPayloadSize`: 1MB (covers `TokenOnLedger` and `TokenOnLedgerMetadata`)
    - `MaxTokenOutputsPerTx`: 1000 outputs
    - `MaxBulkDeleteSize`: 10,000 IDs
    - `MaxWalletIDSize`: 1024 bytes
    - `MaxOwnerRawSize`: 2048 bytes
    - `MaxIssuerRawSize`: 2048 bytes
- **AppendRequest**: Refactored the internal data flow to use a request-based structure for validated processing.
- **Strict Validation Logic**: Implemented `validateAppendRequest` which performs in-memory checks *before* any database transactions or storage operations are initiated.
- **Enhanced Testing**: Added a comprehensive test suite in `token/services/tokens/validation_test.go` covering edge cases, limit violations, and structural validation.

## Verification Results
- **Unit Tests**: All tests in `token/services/tokens` passed successfully.
- **Linting**: Code is clean according to `golangci-lint` (v2) standards, including `nlreturn` and `testifylint` compliance.

Fixes: #1608

### CODE REVIEW

**Comment by HayimShaul** [comment]
@SurbhiAgarwal1 , Thank you for the PR.
I think the PR implements validation in the wrong place. While the validation logic itself is good, it only protects the local token database after transactions are already committed to the ledger.

The primary validation must occur in the Validator before transactions are committed. The current implementation can be kept as a secondary defense layer, but the main mitigation is missing.

Can you please revise?

File: plan.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Implementation Plan - Strict Input Validation in tokens.Service.Append
+
```
Comment (HayimShaul) [inline plan.md]: I don't think this file should be a part of the commit

File: cmd/tokengen/testdata/zkatdlognoghv1_pp.json [hunk 2]
```diff
@@ -1 +1 @@
...JCZ05WQkFnVENrTmhiR2xtYjNKdWFXRXhGakFVQmdOVkJBY1REVk5oYmlCRwpjbUZ1WTJselkyOHhIakFjQmdOVkJBb1RGVTl5WjJsemMzVmxjaTVsZUdGdGNHeGxMbU52YlRFaE1COEdBMVVFCkF4TVlZMkV1VDNKbmFYTnpkV1Z5TG1WNFlXMXdiR1V1WTI5dE1CNFhEVEl5TURjeE9ERTFNRE13TUZvWERUTXkKTURjeE5URTFNRE13TUZvd1lURUxNQWtHQTFVRUJoTUNWVk14RXpBUkJnTlZCQWdUQ2tOaGJHbG1iM0p1YVdFeApGakFVQmdOVkJBY1REVk5oYmlCR2NtRnVZMmx6WTI4eEpUQWpCZ05WQkFNVEhHbHpjM1ZsY2k1UGNtZHBjM04xClpYSXVaWGhoYlhCc1pTNWpiMjB3V1RBVEJnY3Foa2pPUFFJQkJnZ3Foa2pPUFFNQkJ3TkNBQVJacTJZck5ZZGsKbFFjenhMR2ZIdU4yVUxROHExTXovejZKRnBSQ0xUZmdrZEpHTldUU3NWdm9iczlENW10QmlxNDJabTdOTXhNLwplSU9vZDVnZXQ1Ni9vNEhKTUlIR01BNEdBMVVkRHdFQi93UUVBd0lIZ0RBTUJnTlZIUk1CQWY4RUFqQUFNQ3NHCkExVWRJd1FrTUNLQUlGR0s3dXc0NUR6MU4yTGl6YUlDSFFrOWo5Tk5RWS9RdUVCR25GRHo1YStXTUhrR0NDb0QKQkFVR0J3Z0JCRzE3SW1GMGRISnpJanA3SW1obUxrRm1abWxzYVdGMGFXOXVJam9pSWl3aWFHWXVSVzV5YjJ4cwpiV1Z1ZEVsRUlqb2lhWE56ZFdWeUxrOXlaMmx6YzNWbGNpNWxlR0Z0Y0d4bExtTnZiU0lzSW1obUxsUjVjR1VpCk9pSWlMQ0p5Wld4aGVTSTZJbVpoYkhObEluMTlNQW9HQ0NxR1NNNDlCQU1DQTBZQU1FTUNJRHROamdsa29EY0sKaFRjWVFIcTA1V2lrdy9jZGdhRTJtalBGTEE5T211bWdBaDk2bGI5cHBIWEM3WjNGYVNNWGVhUTBkOE9xc0VabgpVWWQ3aEFmeWlrMGUKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQpI////////////AVBAWggKAmsxEgJmMVoICgJrMhICZjI="}
```
Comment (HayimShaul) [inline cmd/tokengen/testdata/zkatdlognoghv1_pp.json:1]: why was this file changed?

**Comment by HayimShaul** [comment]
Can you please also add size limit in protobuf?

File: token/services/tokens/tokens.go [hunk 3]
```diff
@@ -85,10 +101,26 @@ type Service struct {
...
+		NetworkProvider: networkProvider,
+		Storage:         storage,
+		RequestsCache:   requestsCache,
+		Config: ValidationConfig{
+			MaxTokenPayloadSize:  2 * 1024 * 1024, // 2MB
+			MaxTokenOutputsPerTx: 1000,
+			MaxBulkDeleteSize:    10000,
+			MaxWalletIDSize:      1024,
+			MaxOwnerRawSize:      16 * 1024, // 16KB for Idemix
+			MaxIssuerRawSize:     16 * 1024,
```
Comment (HayimShaul) [inline token/services/tokens/tokens.go]: Maybe have these value configurable?

**Comment by adecaro** [comment]
Hi @SurbhiAgarwal1 , any update on this? Any clarification needed? Many thanks 🙏

**Comment by SurbhiAgarwal1** [comment]
Thanks @HayimShaul and @adecaro for the detailed feedback.

You're right — the current validation in `tokens.Service.Append` mainly protects the local token database after the transaction processing flow, so it should not be the primary enforcement point. I’ll revise the implementation to add the main validation earlier in the Validator path before transaction commitment, while keeping the current checks in `tokens.Service.Append` as an additional defense layer.

I’ll also:

* remove unrelated file changes from the PR,
* add protobuf/message size bounds where applicable,
* make validation limits configurable,
* and resolve the current merge conflicts.

I’ll push an updated revision soon. Thanks again for the guidance 🙏

**Comment by SurbhiAgarwal1** [comment]
Hi @HayimShaul and @adecaro ,

I've just pushed a clean, rebased version of this PR. I apologize for the noise earlier—I wanted to make sure the implementation was fully synchronized with the latest upstream changes (especially the recent FSC updates) and that all local regressions were addressed.
